### PR TITLE
Implementation of SocketCAN for Scapy

### PR DESCRIPTION
**Pull request by polybassa**
Clean git history.

### CODE REVIEW

File: doc/scapy/automotive_usage.rst [hunk 1]
```diff
@@ -0,0 +1,387 @@
+****************
+Automotive usage
+****************
+
+.. note::
+    All automotive related features are only supported on linux systems, yet. CAN and ISOTP sockets in scapy3k are based on linux kernel modules.
```
Comment (gpotter2) [inline doc/scapy/automotive_usage.rst]: Hi !
When you copy/paste code, please check what you are doing :/
We're not in `scapy3k`

File: scapy/layers/can.py [hunk 2]
```diff
@@ -1,44 +1,49 @@
...
 from scapy.layers.l2 import CookedLinux
 
-
 # Mimics the Wireshark CAN dissector parameter 'Byte-swap the CAN ID/flags field'
 #   set to True when working with PF_CAN sockets
 conf.contribs['CAN'] = {'swap-bytes': False}
 
+############
+## Consts ##
+############
```
Comment (gpotter2) [inline scapy/layers/can.py]: We don't need this banner

File: scapy/layers/can.py [hunk 3]
```diff
@@ -1,44 +1,49 @@
...
+CAN_FRAME_SIZE = 16
+CAN_INV_FILTER = 0x20000000
 
-class CAN(Packet):
-    """A minimal implementation of the CANopen protocol, based on
-    Wireshark dissectors. See https://wiki.wireshark.org/CANopen
```
Comment (gpotter2) [inline scapy/layers/can.py]: Please leave the comment

File: scapy/layers/can.py [hunk 4]
```diff
@@ -76,6 +81,5 @@ def self_build(self, field_pos_list=None):
             return struct.pack('<I12s', *struct.unpack('>I12s', p))
         return p
 
-
```
Comment (gpotter2) [inline scapy/layers/can.py]: Please don't do unnecessary changes

File: test/can.uts [hunk 5]
```diff
@@ -60,7 +60,6 @@ set(pkt.flags for pkt in packets) == {0}
 
 set(pkt.length for pkt in packets) == {1, 2, 8}
 
-
```
Comment (gpotter2) [inline test/can.uts]: Unnecessary

File: test/cansocket.uts [hunk 6]
```diff
@@ -0,0 +1,251 @@
+% Regression tests for the CANSocket
```
Comment (gpotter2) [inline test/cansocket.uts]: You shall set `~ needs_root` here

Comment (guedou) [inline test/cansocket.uts]: These tests are wrong. They should not try to send packets to the network as it won't work on our build systems. Please either mock the tests, or implement dissection / build as done here https://github.com/secdev/scapy/blob/master/scapy/contrib/gtp.uts

Comment (gpotter2) [reply test/cansocket.uts]: Are you sure of that ?
He managed to make the tests work with sudo enabled, and they look quite OK to me :/
He's performing real CAN socket tests, which shows that the implementation is correct

**Comment by polybassa** [comment]
Thanks for all your comments. I’ll fix them soon. 

If I can install can-utils on the build system, I should be able to use virtual CAN interfaces. I need a hint, how to enable this unit test only on Linux machines, since Travis is testing on all kind of platforms.

File: test/cansocket.uts (hunk 6 above)
Comment (guedou) [reply test/cansocket.uts]: they are all disabled in `test/config/`

Comment (gpotter2) [reply test/cansocket.uts]: Only the Windows ones

**Comment by guedou** [comment]
@polybassa please implement the basic build / dissection tests as I mentioned in my previous comment.
Leave the "live" tests for another PR.

File: test/cansocket.uts (hunk 6 above)
Comment (guedou) [reply test/cansocket.uts]: They won't work on Linux too.

Only testing the code by sending and receiving is a bad idea. Building/dissecting is way better, for example we can be sure that the packets can be read from PCAPs on all platforms.

Comment (gpotter2) [reply test/cansocket.uts]: Got it

**Comment by polybassa** [comment]
Hi,
Thanks for your comments.
 I've removed the cansocket unit test for now. I'll create a new PR with just this unit test.
Nils

File: doc/scapy/automotive_usage.rst [hunk 7]
```diff
@@ -0,0 +1,387 @@
+****************
+Automotive usage
+****************
+
+.. note::
+    All automotive related features are only supported on linux systems, yet. CAN and ISOTP sockets in scapy are based on linux kernel modules.
+    This guide explains the hardware setup on a BeagleBone Black. The BeagleBone Black was chosen because of its two CAN interfaces on the main processor.
+    The presence of two CAN interfaces in one device gives the possibility of CAN MITM attacks and session hijacking.
+    The Cannelloni framework turns a BeagleBone Black into a CAN-to-UDP interface, which gives you the freedom to run Scapy3k
```
Comment (gpotter2) [inline doc/scapy/automotive_usage.rst]: scapy3k

Comment (polybassa) [reply doc/scapy/automotive_usage.rst]: sorry

**Comment by guedou** [comment]
@polybassa my comment is about writing unit tests. This PR won't be merged if you don't write them as explained previously.

**Comment by polybassa** [comment]
Hi @guedou, this PR consists of minor modifications to the CAN Packet implementation and the implementation of a SuperSocket object for CAN. The CAN Packet implementation is already fully tested with dissect and build tests. This can.uts are already part of mainline scapy. I double checked the implementations of the present unit test. There is only one other implementation of a unit test, where a socket object is tested (sendsniff.uts). This unit test is quite similar to my cansocket.uts. I see no other way, to test my socket implementation. Unfortunatly, the support of cansockets is limited to python3 and Linux. The implementation of CAN is platform independant. 
What further steps do I have to do, to make this PR mergeable?
Greetings Nils

**Comment by guedou** [comment]
The idea between this PR is really nice but it needs some fix:
- do not change can.py in this PR. If something is wrong in this file, please open another independent PR for discussions
- as it is Linux dependent, the code from cansocket.py must be moved in scapy/arch/linux.py Make sure that it can not be loaded by another architecture
- please use a linter such as pylint, there are some problems easy to fix

Regarding the Python3 only support. I would prefer that you provide something that is also compatile with Python 2.7, but I can leave without.

Have you tried passing the CAN_RAW constant as an integer in Python 2.7 ?  Have you tried using ctypes to directly call socket from libc ? See https://github.com/secdev/scapy/blob/master/scapy/arch/bpf/core.py#L21 for an example with ioctl

**Comment by polybassa** [comment]
Hi @guedou, 
for Python 2.7: It is possible to use ctypes to create the socket. This was done by this project: https://github.com/hardbyte/python-can/blob/develop/can/interfaces/socketcan/socketcan_ctypes.py
Unfortunatly, I hadn't enough time yet, to provide a clean port of this code for scapy. But I can do this at some point in the future. Using the built-in support of python3 was easier, in the past. To proceed with the CAN support for scapy, I suggest the following steps:

- PR: Minor changes to can.py. After this PR is in -->
- PR: CANSocket implementation for scapy (python3 only) AND
- ISSUE: (for me) add CANSocket implementation for Python 2.7
- PR: ISOTP, UDS and DoIP Support for Scapy

In the future, I would like also to implement GMLAN and some more automotive protocols. 


I will put SuperSocket implementations into scapy/arch/linux, and packet/protocol implementations into scapy/, right?

Greetings Nils
